OpenSim/Common/FileSystemPath.h

```diff
@@ -0,0 +1,122 @@
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+// INCLUDES
+#include <OpenSim/Common/osimCommonDLL.h>
+#include "OpenSim/Common/IO.h"
+#include "OpenSim/Common/Object.h"
+#include "Simbody.h"
+
+namespace OpenSim {
+
+//==============================================================================
+//                            OPENSIM FILESYSTEMPATH
+//==============================================================================
+/**
+* A class for handling file system paths. The class holds two variables: 1) a
+* string for the path and 2) a bool _isFile to denote of the path is a file
+* or a directory.
+*
+* For convention, the word "Path" can refer to the FileSystemPath for either
+* a file or directory. "FilePath" and "DirPath" are specific to a file or
+* directory, respectively.
+*
+* This class has methods for resolving paths with respect to some other given
+* directory (i.e. not necessarily the current working directory). In order to
+* resolve these cases, we use methods in SimTK::Pathname. Refer to that class's
+* documentation for the rules of how these are evaluated.
+*
+* @author Carmichael Ong
+*/
+class FileSystemPath : public Object {
+    OpenSim_DECLARE_CONCRETE_OBJECT(FileSystemPath, Object);
+
+public:
+    /** Default constructor **/
+    FileSystemPath();
+
+    /** Construct FileSystemPath from a string. **/
+    FileSystemPath(std::string& pathName);
+    FileSystemPath(std::string& pathName, bool isFile);
+
+    /** Use default copy constructor and assignment operator. */
+    FileSystemPath(const FileSystemPath&) = default;
+    FileSystemPath& operator=(const FileSystemPath&) = default;
+
+    /** Destructor. **/
+    ~FileSystemPath() = default;
+
+    // Get a FileSystemPath for the current working directory.
+    FileSystemPath getWorkingDirPath();
+
+    // Get absolute path to a file or directory. getAbsoluteFilePath() and
+    // getAbsoluteDirPath() checks if the FileSystemPath is a file or dir
+    // first. getAbsolutePath() checks if the FileSystemPath is a file or dir
+    // then calls the corresponding function. If the path was not an absolute 
+    // path, then the directory is resolved with respect to the current working 
+    // directory.
+    FileSystemPath getAbsolutePath();
+    FileSystemPath getAbsoluteFilePath();
+    FileSystemPath getAbsoluteDirPath();
+
+    // Get the absolute path for the directory of a file. Checks if it is
+    // given a file first, then gives the absolute path to the directory
+    // containing the file. If the path was not an absolute path, then the
+    // directory is resolved with respect to the current working directory.
+    FileSystemPath getAbsoluteDirPathForFile();
+
+    // Return the absolute path relative to some specified relativeDir.
+    FileSystemPath getAbsolutePathWithRelativeDir(FileSystemPath relativeDir);
```

Comment:
take the argument by const reference. This just appends two paths together? This is a little confusing (to say "absolute path relative to x"), but I think I get what you're saying. If I have a path "x/y/z" and I know it is relative to "u/v", then what you want is "abs(pwd)/u/v/x/y/z".?